### DISCUSSION TITLE
Is it by-design or a bug that button icon's color doesn't change with its text?

### DISCUSSION THREAD

**Discussion by datvm** [material-components/material-web#5160] [Q&A] [answered]
[Lit Playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgXCJAbWF0ZXJpYWwvd2ViL2FsbFwiIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8bGluayByZWw9XCJzdHlsZXNoZWV0XCIgaHJlZj1cImh0dHBzOi8vZm9udHMuZ29vZ2xlYXBpcy5jb20vY3NzMj9mYW1pbHk9TWF0ZXJpYWwrU3ltYm9scytPdXRsaW5lZDpvcHN6LHdnaHQsRklMTCxHUkFEQDIwLi40OCwxMDAuLjcwMCwwLi4xLC01MC4uMjAwXCIgLz5cbiAgXG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPG1kLXRleHQtYnV0dG9uIHN0eWxlPVwiLS1tZC10ZXh0LWJ1dHRvbi1sYWJlbC10ZXh0LWNvbG9yOiByZWRcIj5cbiAgICA8bWQtaWNvbiBzbG90PVwiaWNvblwiPnNlbmQ8L21kLWljb24-XG4gICAgSGVsbG9cbiAgPC9tZC10ZXh0LWJ1dHRvbj5cbiAgPG1kLWZpbGxlZC1idXR0b24gc3R5bGU9XCItLW1kLWZpbGxlZC1idXR0b24tbGFiZWwtdGV4dC1jb2xvcjogY3lhblwiPlxuICAgIDxtZC1pY29uIHNsb3Q9XCJpY29uXCI-c2VuZDwvbWQtaWNvbj5cbiAgICBIZWxsb1xuICA8L21kLWZpbGxlZC1idXR0b24-XG4gIDxtZC1maWxsZWQtYnV0dG9uIHN0eWxlPVwiLS1tZC1maWxsZWQtYnV0dG9uLWxhYmVsLXRleHQtY29sb3I6IGN5YW47IC0tX2ljb24tY29sb3I6IGN5YW47XCI-XG4gICAgPG1kLWljb24gc2xvdD1cImljb25cIj5zZW5kPC9tZC1pY29uPlxuICAgIEhlbGxvXG4gIDwvbWQtZmlsbGVkLWJ1dHRvbj5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4zLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl40LjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMy4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0):

![image](https://github.com/material-components/material-web/assets/6388546/d4033baa-4c6d-4b3a-bd66-72d2f0fed648)

```html
  <md-text-button style="--md-text-button-label-text-color: red">
    <md-icon slot="icon">send</md-icon>
    Hello
  </md-text-button>
  <md-filled-button style="--md-filled-button-label-text-color: cyan">
    <md-icon slot="icon">send</md-icon>
    Hello
  </md-filled-button>
  <md-filled-button style="--md-filled-button-label-text-color: cyan; --_icon-color: cyan;">
    <md-icon slot="icon">send</md-icon>
    Hello
  </md-filled-button>
```

I also noticed when hovering the button, the color revert to default and all the variables start with a `_` so I assume it's private and isn't meant to be modified?

**Comment by asyncliz** [comment, marked as answer]
It is by design, since label text and icons have different tokens.

```css
--md-text-button-label-text-color: red;
--md-text-button-icon-color: red;
```

> You're correct that the `--_underscore-tokens` are not meant to be used. However, they can be a good indication of what the name is. `--_icon-color` means `--md-text-button-icon-color`, `--md-filled-button-icon-color`, and so on.

Many tokens also have additional tokens for states like hover, pressed, focus, and disabled that need to be set, which is why the color reverts on hover.

```css
--md-text-button-label-text-color: red;
--md-text-button-hover-label-text-color: red;
--md-text-button-focus-label-text-color: red;
--md-text-button-pressed-label-text-color: red;
```

You can see a list of supported tokens for each variant in the [theming section of the component docs](https://github.com/material-components/material-web/blob/main/docs/components/button.md#theming). For example, here's the link to the [elevated button's tokens](https://github.com/material-components/material-web/blob/7d7858090e1ad5b128d4e20c036e2312ae460788/tokens/_md-comp-elevated-button.scss#L19-L60).

We acknowledge this is a lot, and we hope to improve things 😣

One recommendation we suggest is to use `--md-sys-*` tokens to help with styling instead of component tokens. For example, buttons use the `primary` color, so we can change the overall look without needing to override every token:

```css
.cyan-button {
  --md-sys-color-primary: cyan;
  --md-sys-color-on-primary: black;
}
```

Check out [How do I change the color of a button?](https://github.com/material-components/material-web/blob/main/docs/support.md#how-do-i-change-the-color-of-a-button)

**Reply by datvm** [reply]
Ah clever, I didn't see that part of the docs. Will be using `--md-sys-color-primary` and `--md-sys-color-on-primary` variables from now on, thanks!
